src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+
+                case State.ExceptionNext:
+                    {
+                        var current = _cursor.GetNextFilter<IExceptionFilter, IAsyncExceptionFilter>();
+                        if (current.FilterAsync != null)
+                        {
+                            state = current.FilterAsync;
+                            goto case State.ExceptionAsyncBegin;
+                        }
+                        else if (current.Filter != null)
+                        {
+                            state = current.Filter;
+                            goto case State.ExceptionSyncBegin;
+                        }
+                        else if (scope == Scope.Exception)
+                        {
+                            goto case State.ExceptionInside;
+                        }
+                        else
+                        {
+                            Debug.Assert(scope == Scope.Invoker || scope == Scope.Resource);
+                            goto case State.ActionBegin;
+                        }
+                    }
+
+                case State.ExceptionAsyncBegin:
+                    {
+                        var task = InvokeNextExceptionFilterAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ExceptionAsyncResume;
+                            return task;
+                        }
+
+                        goto case State.ExceptionAsyncResume;
+                    }
+
+                case State.ExceptionAsyncResume:
+                    {
+                        var filter = (IAsyncExceptionFilter)state;
+                        var exceptionContext = _exceptionContext;
+
+                        if (exceptionContext?.Exception != null)
+                        {
+                            _diagnosticSource.BeforeOnExceptionAsync(exceptionContext, filter);
+
+                            // Exception filters only run when there's an exception - unsetting it will short-circuit
+                            // other exception filters.
+                            var task = filter.OnExceptionAsync(exceptionContext);
+                            if (task.Status != TaskStatus.RanToCompletion)
+                            {
+                                next = State.ExceptionAsyncEnd;
+                                return task;
+                            }
+
+                            goto case State.ExceptionAsyncEnd;
+                        }
+
+                        goto case State.ExceptionEnd;
+                    }
+
+                case State.ExceptionAsyncEnd:
+                    {
+                        var filter = (IAsyncExceptionFilter)state;
+                        var exceptionContext = _exceptionContext;
+
+                        _diagnosticSource.AfterOnExceptionAsync(exceptionContext, filter);
+
+                        if (exceptionContext.Exception == null)
```

Comment:
exceptionContext can be null here if no exceptions happen, can't it? If so, can you add a test to cover this case?